src/Kunstmaan/NodeBundle/Controller/SlugController.php

```diff
@@ -49,56 +78,88 @@ public function slugAction(Request $request, $url = null, $preview = false)
             throw $this->createNotFoundException('No page found for slug ' . $url);
         }
 
-        $entity = $this->getPageEntity(
-            $request,
-            $preview,
-            $em,
-            $nodeTranslation
-        );
-        $node   = $nodeTranslation->getNode();
+        return $nodeTranslation;
+    }
 
+    /**
+     * @param Node $node
+     * @param EntityInterface $entity
+     * @param Request $request
+     * @param NodeTranslation $nodeTranslation
+     * @return SlugSecurityEvent
+     */
+    private function getSecurityEvent(Node $node, EntityInterface $entity, Request $request, NodeTranslation $nodeTranslation)
+    {
         $securityEvent = new SlugSecurityEvent();
         $securityEvent
             ->setNode($node)
             ->setEntity($entity)
             ->setRequest($request)
             ->setNodeTranslation($nodeTranslation);
 
+        return $securityEvent;
+    }
+
+    /**
+     * @param $locale
+     * @param Node $node
+     * @param bool $preview
+     *
+     * @return NodeMenu
+     */
+    private function getNodeMenu($locale, Node $node, $preview)
+    {
+        /** @var NodeMenu $nodeMenu */
         $nodeMenu = $this->container->get('kunstmaan_node.node_menu');
         $nodeMenu->setLocale($locale);
         $nodeMenu->setCurrentNode($node);
         $nodeMenu->setIncludeOffline($preview);
 
-        $eventDispatcher = $this->get('event_dispatcher');
-        $eventDispatcher->dispatch(Events::SLUG_SECURITY, $securityEvent);
-
-        //render page
-        $renderContext = new RenderContext(
-            array(
-                'nodetranslation' => $nodeTranslation,
-                'slug'            => $url,
-                'page'            => $entity,
-                'resource'        => $entity,
-                'nodemenu'        => $nodeMenu,
-            )
-        );
+        return $nodeMenu;
+    }
+
+    /**
+     * @param NodeTranslation $nodeTranslation
+     * @param $url
+     * @param EntityInterface $entity
+     * @param NodeMenu $nodeMenu
+     * @param SlugSecurityEvent $securityEvent
+     * @return RenderContext
+     */
+    private function getRenderContext(NodeTranslation $nodeTranslation, $url, EntityInterface $entity, NodeMenu $nodeMenu, SlugSecurityEvent $securityEvent)
+    {
+        $this->eventDispatcher = $this->get('event_dispatcher');
```

Comment:
It's strange that the eventdispatcher is collected from container here. Maybe move it to correct place.